loopchain/peermanager/peer_loader.py

```diff
@@ -13,57 +13,53 @@
 # limitations under the License.
 """PeerListData Loader for PeerManager"""
 
-import logging
 import os
-from typing import TYPE_CHECKING
 
 from loopchain import configure as conf
 from loopchain import utils
 from loopchain.baseservice import ObjectManager, RestMethod
+from loopchain.blockchain.types import Hash32
 from loopchain.channel.channel_property import ChannelProperty
-from loopchain.peermanager import Peer
-
-if TYPE_CHECKING:
-    from loopchain.peermanager import PeerManager
 
 
 class PeerLoader:
     def __init__(self):
         pass
 
     @staticmethod
-    def load(peer_manager: 'PeerManager'):
-        if os.path.exists(conf.CHANNEL_MANAGE_DATA_PATH):
-            PeerLoader._load_peers_from_file(peer_manager)
+    def load():
+        peers = PeerLoader._load_peers_from_db()
+        if peers:
+            utils.logger.info("Reps data loaded from DB")
+            return peers
+        elif os.path.exists(conf.CHANNEL_MANAGE_DATA_PATH):
+            utils.logger.info(f"Try to load reps data from {conf.CHANNEL_MANAGE_DATA_PATH}")
+            return PeerLoader._load_peers_from_file()
         else:
-            PeerLoader._load_peers_from_rest_call(peer_manager)
+            utils.logger.info("Try to load reps data from other reps")
+            return PeerLoader._load_peers_from_rest_call()
+
+    @staticmethod
+    def _load_peers_from_db() -> list:
+        blockchain = ObjectManager().channel_service.block_manager.blockchain
+        last_block = blockchain.last_block
+        rep_root_hash = (last_block.header.reps_hash if last_block else
+                         Hash32.fromhex(conf.CHANNEL_OPTION[ChannelProperty().name].get('crep_root_hash')))
+
+        return blockchain.find_preps_by_roothash(rep_root_hash)
 
     @staticmethod
-    def _load_peers_from_file(peer_manager: 'PeerManager'):
-        utils.logger.debug(f"load_peers_from_file")
+    def _load_peers_from_file():
         channel_info = utils.load_json_data(conf.CHANNEL_MANAGE_DATA_PATH)
         reps: list = channel_info[ChannelProperty().name].get("peers")
-        for peer in reps:
-            peer_manager.add_peer(peer)
+        return [{"id": rep["id"], "p2pEndpoint": rep["peer_target"]} for rep in reps]
 
     @staticmethod
-    def _load_peers_from_rest_call(peer_manager: 'PeerManager'):
+    def _load_peers_from_rest_call():
         rs_client = ObjectManager().channel_service.rs_client
         crep_root_hash = conf.CHANNEL_OPTION[ChannelProperty().name].get('crep_root_hash')
-
-        if not crep_root_hash:
```

Comment:
Is this check no longer needed?

Reply:
If there is last block, get reps_hash from the block.
if not, try to search reps_hash by using `crep_root_hash`.
But this situation may not be allowed within decentralized current MainNet, because all nodes has last block (by BlockSync or whatever)